internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+package metrics
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"math"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	"git.sr.ht/~spc/go-log"
+
+	"github.com/golang/snappy"
+	"github.com/project-flotta/flotta-operator/models"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+const (
+	defaultWaitInterval         = 5 * time.Minute
+	defaultRequestNumSamples    = 10000
+	defaultRequestDuration      = time.Hour
+	defaultRequestRetryInterval = 10 * time.Second
+	LastWriteFileName           = "metrics-lastwrite"
+	DeviceLabel                 = "edgedeviceid"
+)
+
+type RemoteWrite struct {
+	deviceID             string
+	config               RemoteWriteConfiguration
+	tsdb                 API
+	LastWrite            time.Time
+	lastWriteFile        string
+	waitInterval         time.Duration
+	RangeDuration        time.Duration
+	RequestNumSamples    int
+	RequestRetryInterval time.Duration
+	Client               WriteClient
+}
+
+type RemoteWriteConfiguration struct {
+	Url     string // empty URL means that feature is disabled
+	Timeout time.Duration
+}
+
+func NewRemoteWrite(dataDir, deviceID string, tsdbInstance API) *RemoteWrite {
+	waitInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_WAIT_INTERVAL"))
+	if waitInterval == 0 {
+		waitInterval = defaultWaitInterval
+	}
+	RangeDuration, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_RANGE_DURATION"))
```

Comment:
Overall, I think that all of those env variables should be part of the main device config, no? so users can update from config.

Reply:
it is internal. not exposed to users and hopefully never will be changed. the env vars are more for development

Reply:
then a const, for dev you can update.

Reply:
The use cases are: 
- If a customer for X reasons has that ENV variable for any reason, the program will miss behaving, and also it's going to be hard to debug. . 
- Those devices can be in any place without any physical security, you don't want to have any env variable out there that it's not documented that can be a vector of attack.

So, if it's not going to be in the deviceConfig from the API, I wouldn't take an extra risk on that.

Reply:
we have a lot of env vars just like these and we never raised these concerns. I think you took it too far.

Reply:
we have a lot of env vars on the operator, not in the device, which is a big difference for me.

Reply:
fixed